pkg/providers/cloudstack/cloudstack_test.go

```diff
@@ -198,6 +198,32 @@ func TestProviderGenerateCAPISpecForCreate(t *testing.T) {
 	test.AssertContentToFile(t, string(md), "testdata/expected_results_main_md.yaml")
 }
 
+func TestProviderGenerateDeploymentFileWithAffinity(t *testing.T) {
```

Comment:
Can we have error cases as well because both can't be specified at the same time? And do we unit tests for group ids also?

Reply:
pkg/providers/cloudstack/testdata/cluster_main.yaml and TestProviderGenerateCAPISpecForCreate() test against affinityGroupIds. 

A unit test to check if validation fails when both are defined is in TestValidateMachineConfigsWithAffinity() 

```
	// Both affinity and affinityGroupIds are defined
	cloudStackClusterSpec.controlPlaneMachineConfig().Spec.AffinityGroupIds = []string{"affinity-group-1"}
	err = validator.ValidateClusterMachineConfigs(ctx, cloudStackClusterSpec)
	assert.NotNil(t, err)
	cloudStackClusterSpec.controlPlaneMachineConfig().Spec.Affinity = originalValue
```